src/hotspot/share/opto/phaseX.cpp

```diff
@@ -2795,43 +2796,84 @@ void PhaseCCP::analyze() {
-    DEBUG_ONLY(worklist_verify.push(n);)
-    if (n->is_SafePoint()) {
-      // Make sure safepoints are processed by PhaseCCP::transform even if they are
-      // not reachable from the bottom. Otherwise, infinite loops would be removed.
-      _root_and_safepoints.push(n);
-    }
-    const Type* new_type = n->Value(this);
-    if (new_type != type(n)) {
-      DEBUG_ONLY(verify_type(n, new_type, type(n));)
-      dump_type_and_node(n, new_type);
-      set_type(n, new_type);
-      push_child_nodes_to_worklist(worklist, n);
-    }
-    if (KillPathsReachableByDeadTypeNode && n->is_Type() && new_type == Type::TOP) {
-      // Keep track of Type nodes to kill CFG paths that use Type
-      // nodes that become dead.
-      _maybe_top_type_nodes.push(n);
+  bool first_round = true;
+  do {
+    // Start the CCP round.
+    while (worklist.size() != 0) {
+      Node* n = fetch_next_node(worklist);
+      if (first_round) {
+        DEBUG_ONLY(worklist_verify.push(n);)
+        if (needs_revisit(n)) {
+          worklist_revisit.push(n);
+        }
+        if (n->is_SafePoint()) {
+          // Make sure safepoints are processed by PhaseCCP::transform even if they are
+          // not reachable from the bottom. Otherwise, infinite loops would be removed.
+          _root_and_safepoints.push(n);
+        }
+      }
+      const Type* new_type = n->Value(this);
+      if (new_type != type(n)) {
+        DEBUG_ONLY(verify_type(n, new_type, type(n));)
+        dump_type_and_node(n, new_type);
+        set_type(n, new_type);
+        push_child_nodes_to_worklist(worklist, n);
+      }
+      if (KillPathsReachableByDeadTypeNode && n->is_Type() && new_type == Type::TOP) {
+        // Keep track of Type nodes to kill CFG paths that use Type
+        // nodes that become dead.
+        _maybe_top_type_nodes.push(n);
+      }
     }
-  }
+
+    // Check if any nodes slated for revisit are able to refine their types at the end of the round.
+    // If so, push them to worklist to start another round and process them.
+    for (uint t = 0; t < worklist_revisit.size(); t++) {
+      Node* n = worklist_revisit.at(t);
+      const Type* new_type = n->Value(this);
+      if (new_type != type(n)) {
+        worklist.push(n);
+      }
+    }
+
+    // First round is done.
+    first_round = false;
+  } while (worklist.size() != 0);
+
   DEBUG_ONLY(verify_analyze(worklist_verify);)
 }
 
+// Some nodes can refine their types due to type change somewhere deep
+// in the graph. We will need to revisit them before claiming convergence.
+// Add nodes here if particular *Node::Value is doing deep graph traversals
+// not handled by push_more_uses.
+bool PhaseCCP::needs_revisit(Node *n) const {
+  // LoadNode performs deep traversals. Load is not notified for changes far away.
+  if (n->is_Load()) {
+    return true;
+  }
+  // CmpPNode performs deep traversals if it compares oopptr. CmpP is not notified for changes far away.
+  if (n->Opcode() == Op_CmpP) {
```

Comment:
The verification restricts it to `n->Opcode() == Op_CmpP && type(n->in(1))->isa_oopptr() && type(n->in(2))->isa_oopptr()`. How big is the difference here? Might this have a performance impact?

Reply:
Honestly, no idea. I just wanted to have a conservative check, e.g. "We know `CmpP` does something fishy? We are going to revisit it." But it will make sense to keep C2 compilation fast. Let me try to add the oopptr checks and see if anything shows up in CTW.

Reply:
Seems to work fine on large CTW corpus run, so I am leaving this additional check in. Thanks!